##Project for a B2B professional services company

## Task 1
#### Has user activity or engagement dropped, increased or remained stable? What is the extent of change in user activity/engagement?

##### First let us check the duration of the dataset

In [0]:
%sql 
SELECT min(occurred_at), max(occurred_at) FROM events

min(occurred_at) max(occurred_at) 2014-05-01 00:54:09 2014-08-31 20:40:32

##### Since the data is for four months (from May to August), let us plot total user engagement per day, week and month for the entire dataset

In [0]:
%sql
-- by day
SELECT
   to_date(occurred_at) as date,
   COUNT (*) as total_engagement 
FROM
   events
WHERE
   event_type = 'engagement' 
GROUP BY
   to_date(occurred_at) 
ORDER BY
   to_date(occurred_at)

date total_engagement 2014-05-01 1503 2014-05-02 1969 2014-05-03 442 2014-05-04 271 2014-05-05 1200 2014-05-06 1494 2014-05-07 1361 2014-05-08 1496 2014-05-09 1468 2014-05-10 449 2014-05-11 319 2014-05-12 1025 2014-05-13 1395 2014-05-14 1620 2014-05-15 1572 2014-05-16 1660 2014-05-17 486 2014-05-18 221 2014-05-19 1220 2014-05-20 1377 2014-05-21 1373 2014-05-22 1484 2014-05-23 1515 2014-05-24 558 2014-05-25 304 2014-05-26 1129 2014-05-27 1366 2014-05-28 1344 2014-05-29 1499 2014-05-30 1509 2014-05-31 449 2014-06-01 294 2014-06-02 1240 2014-06-03 1287 2014-06-04 1714 2014-06-05 1402 2014-06-06 1676 2014-06-07 519 2014-06-08 295 2014-06-09 994 2014-06-10 1299 2014-06-11 1356 2014-06-12 1504 2014-06-13 1482 2014-06-14 532 2014-06-15 351 2014-06-16 1281 2014-06-17 1457 2014-06-18 1534 2014-06-19 1405 2014-06-20 1461 2014-06-21 637 2014-06-22 183 2014-06-23 1225 2014-06-24 1372 2014-06-25 1171 2014-06-26 1454 2014-06-27 1569 2014-06-28 466 2014-06-29 382 2014-06-30 1222 2014-07-01 1386 2014-07-02 1373 2014-07-03 1566 2014-07-04 1657 2014-07-05 583 2014-07-06 216 2014-07-07 1274 2014-07-08 1377 2014-07-09 1404 2014-07-10 1393 2014-07-11 1402 2014-07-12 574 2014-07-13 355 2014-07-14 1314 2014-07-15 1387 2014-07-16 1512 2014-07-17 1265 2014-07-18 1805 2014-07-19 555 2014-07-20 232 2014-07-21 1112 2014-07-22 1195 2014-07-23 1501 2014-07-24 1439 2014-07-25 1516 2014-07-26 418 2014-07-27 410 2014-07-28 1322 2014-07-29 1471 2014-07-30 1498 2014-07-31 1543 2014-08-01 1540 2014-08-02 614 2014-08-03 303 2014-08-04 829 2014-08-05 1136 2014-08-06 1236 2014-08-07 1324 2014-08-08 1268 2014-08-09 473 2014-08-10 159 2014-08-11 648 2014-08-12 1021 2014-08-13 1004 2014-08-14 1309 2014-08-15 1309 2014-08-16 428 2014-08-17 203 2014-08-18 972 2014-08-19 910 2014-08-20 1225 2014-08-21 1151 2014-08-22 1143 2014-08-23 417 2014-08-24 186 2014-08-25 914 2014-08-26 1022 2014-08-27 1011 2014-08-28 1215 2014-08-29 1173 2014-08-30 458 2014-08-31 245

In [0]:
%sql
-- by week
SELECT
   EXTRACT(week from occurred_at) as week,
   COUNT (*) as total_engagement 
FROM
   events
WHERE
   event_type = 'engagement' 
GROUP BY
   week 
ORDER BY
   week

week total_engagement 18 4185 19 7787 20 7979 21 7831 22 7590 23 8133 24 7518 25 7958 26 7639 27 8003 28 7779 29 8070 30 7591 31 8291 32 6425 33 5922 34 6004 35 6038

In [0]:
%sql
-- by month
SELECT
   EXTRACT(month from occurred_at) as month,
   COUNT (*) as total_engagement 
FROM
   events 
WHERE
   event_type = 'engagement' 
GROUP BY
   month 
ORDER BY
   month

month total_engagement 5 35078 6 32764 7 36055 8 26846

#### Few Initial Observations
  - We can see that there is a decline is user engagement over the weekends, which is expected because very few people would be working over the weekend
  - We also see a visible decline in user engagement starting from August

#### Given the above obervations, let us analyze the engagement difference week over week and month over month

In [0]:
%sql
-- Analyzing engagement difference by week
WITH weekly_engagement AS 
(
   SELECT
      EXTRACT(week from occurred_at) AS week,
      COUNT (*) as total_engagement 
   FROM
      events
   WHERE
      event_type = 'engagement' 
     AND occurred_at > '2014-05-05' 
   GROUP BY
      1 
   ORDER BY
      1
),
engagement_previous AS 
(
   SELECT
      week,
      total_engagement,
      LAG(total_engagement, 1) OVER ( 
   ORDER BY
      week ) previous_week_engagement 
   FROM
      weekly_engagement 
)
SELECT
   week,
   total_engagement,
   CASE WHEN previous_week_engagement IS NULL THEN 0 ELSE 100*(total_engagement - previous_week_engagement) / previous_week_engagement END as engagement_difference 
FROM
   engagement_previous

week total_engagement engagement_difference 19 7787 0.0 20 7979 2.4656478746628996 21 7831 -1.8548690312069183 22 7590 -3.077512450517175 23 8133 7.154150197628459 24 7518 -7.561785319070454 25 7958 5.852620377760043 26 7639 -4.008544860517718 27 8003 4.765021599685823 28 7779 -2.798950393602399 29 8070 3.740840725028924 30 7591 -5.935563816604708 31 8291 9.221446449743118 32 6425 -22.506332167410445 33 5922 -7.828793774319066 34 6004 1.3846673421141507 35 6038 0.5662891405729513

- We can see that the engagement dropped by almost 22% week over week in the first week of August which seems like an anomaly compared to the rest of the weeks

In [0]:
%sql
-- Analyzing engagement difference month over month
WITH monthly_engagement AS 
(
   SELECT
      EXTRACT(month from occurred_at) AS month,
      COUNT (*) as total_engagement 
   FROM
      events 
   WHERE
      event_type = 'engagement'  
   GROUP BY
      1 
   ORDER BY
      1
),
engagement_previous AS 
(
   SELECT
      month,
      total_engagement,
      LAG(total_engagement, 1) OVER ( 
   ORDER BY
      month ) previous_month_engagement 
   FROM
      monthly_engagement 
)
SELECT
   month,
   total_engagement,
   CASE WHEN previous_month_engagement IS NULL THEN 0 ELSE 100*(total_engagement - previous_month_engagement) / previous_month_engagement END as engagement_difference 
FROM
   engagement_previous

month total_engagement engagement_difference 5 35078 0.0 6 32764 -6.5967272934602885 7 36055 10.044561103650349 8 26846 -25.541533767854666

#### Final Findings

- User activity saw a sharp decline in the month of August (by almost 25%)
- User activity remained almost stable for first three months with a slight drop in the second month
- The extent of change was normal for the first three months while for last two months the extent of change was quite sharp

## Task 2
#### Are there any changes in the three stages of the emails funnel? What changes have you discovered? (Please search online for terms such as sales funnel, marketing funnel to understand the funnel concept).

#### ANSWER FOR TASK 2
##### First let us plot the count of each email action

In [0]:
%sql
SELECT
   action,
   COUNT(*) as num_events 
FROM
   emails 
GROUP BY
   action 
ORDER BY
   num_events DESC

action num_events sent_weekly_digest 57267 email_open 20459 email_clickthrough 9010 sent_reengagement_email 3653

##### Now, let us analyze the drop in user actions for each funnel

In [0]:
%sql
WITH email_actions AS 
(
   SELECT
      action,
      COUNT(*) as num_events 
   FROM
      emails 
   GROUP BY
      action 
   ORDER BY
      num_events DESC 
)
,
funnel AS 
(
   SELECT
      action,
      num_events,
      LAG(num_events, 1) OVER ( 
   ORDER BY
      num_events DESC) previous_action_events 
   FROM
      email_actions 
)
SELECT
   action,
   num_events,
   100*num_events / (
   SELECT
      MAX(num_events) 
   FROM
      funnel) AS overall_conversion,
      100*(num_events) / previous_action_events AS stage_conversion 
   FROM
      funnel

action num_events overall_conversion stage_conversion sent_weekly_digest 57267 100.0 null email_open 20459 35.72563605566906 35.72563605566906 email_clickthrough 9010 15.7333193636824 44.03929810841195 sent_reengagement_email 3653 6.378891857439712 40.54384017758046

#### Findings
- The percentage change in the change in the email funnel is seen in the table above
- The overall_conversation shows the percentage change with respect to the sent_weekly_digest while the stage_conversation shows the percentage change with respect to the previous stage of the funnel
- From the data we can say that if the email is sent to 100 people then around 35% of them opens the email while 15% of the people clicks on the links followed by 6% of the users sent the reengagement email. 
- So yes there is lot of changes in the three stages of the funnel

## Task 3

#### Are changes in user activity/engagement associated with specific devices or hardware ( mobile phone, tablet computers, desktop computers, etc.)  that customers use to read emails and interact with Zilto’sweb portal?

In [0]:
%sql
WITH emails AS
(
   SELECT
      *,
      to_date(occurred_at) AS date,
      EXTRACT(month FROM occurred_at) AS month 
   FROM
      emails emails
)
,
events AS
(
   SELECT
      user_id,
      to_date(occurred_at) AS date,
      device,
      EXTRACT(month FROM occurred_at) AS month 
   FROM
      events
   ORDER BY
      user_id ASC
)
SELECT
   CASE
      WHEN
         device IN 
         (
            'amazon fire phone',
            'nexus 10',
            'iphone 5',
            'nexus 7',
            'iphone 5s',
            'nexus 5',
            'htc one',
            'iphone 4s',
            'samsung galaxy note',
            'nokia lumia 635',
            'samsung galaxy s4'
         )
      THEN
         'Mobile' 
      WHEN
         device IN 
         (
            'ipad mini', 'samsung galaxy tablet', 'kindle fire', 'ipad air'
         )
      THEN
         'Tablet' 
      WHEN
         device IN 
         (
            'dell inspiron notebook', 'macbook pro', 'asus chromebook', 'windows surface', 'macbook air', 'lenovo thinkpad', 'mac mini', 'acer aspire notebook'
         )
      THEN
         'Laptops' 
      WHEN
         device IN 
         (
            'dell inspiron desktop', 'acer aspire desktop', 'hp pavilion desktop'
         )
      THEN
         'Desktops' 
      ELSE
         'Unknown' 
   END
   AS device_type, 
   emails.month, 
   count(emails.user_id) 
FROM
   emails 
   LEFT JOIN
      events 
      ON emails.user_id = events.user_id 
      AND emails.date = events.date 
WHERE
   action = 'email_clickthrough' 
GROUP BY
   1, 2
ORDER BY
   1, 2

device_type month count(user_id) Desktops 5 306 Desktops 6 505 Desktops 7 454 Desktops 8 323 Laptops 5 3587 Laptops 6 3183 Laptops 7 4103 Laptops 8 3002 Mobile 5 2236 Mobile 6 2556 Mobile 7 2579 Mobile 8 246 Tablet 5 460 Tablet 6 427 Tablet 7 444 Tablet 8 34 Unknown 5 1411 Unknown 6 1593 Unknown 7 2026 Unknown 8 1599

#### Findings

##### Findings : 
- From the graph we can say that the

## Task 4
#### Are changes in user activity/engagement associated with specific countries or regions of the world where Zilto’s  users are located?

#### ANSWER FOR TASK 4
##### First let us analyze the number of countries/location present in the data

In [0]:
%sql
SELECT COUNT(DISTINCT(location)) FROM events

count(DISTINCT location) 47

##### Let us plot the top 20 countries which have highest engagement overall

In [0]:
%sql
SELECT location, COUNT(*) as engagement_count FROM events GROUP BY location ORDER BY engagement_count DESC LIMIT 20

location engagement_count United States 42896 Japan 11729 Germany 10286 United Kingdom 6796 France 6783 Russia 6348 Italy 4970 Brazil 4555 India 4108 Canada 3919 Mexico 3553 Korea 3551 Spain 3134 Australia 3050 Sweden 2218 Saudi Arabia 2167 Indonesia 1930 Netherlands 1849 Taiwan 1760 Poland 1672

### Final Findings
- From the bar graph we can see the top 20 countries where the user engagement is highest
- We can say that the user engagement is hightest in the USA which is more than 40k
- Following by Japan and Germany which contributes only around 1/3rd of the US users
- So yes we can say that the activity/engagement is accociated with the country

In [0]:
%sql
SELECT location, COUNT(*) as engagement_count FROM events GROUP BY location ORDER BY engagement_count ASC LIMIT 27

location engagement_count Portugal 265 Pakistan 332 Iraq 375 Argentina 545 Ireland 545 Singapore 570 Hong Kong 572 Thailand 607 Philippines 635 Greece 672 Chile 690 United Arab Emirates 701 Turkey 788 Venezuela 819 Israel 832 Nigeria 848 Colombia 873 Finland 1044 Malaysia 1044 South Africa 1058 Switzerland 1081 Denmark 1133 Norway 1241 Austria 1252 Egypt 1275 Iran 1414 Belgium 1515

### Findings
- In the above graph we can see the 27 countries where the user engagement was the lowest
- We can say that Portugal has the least user engagement followed by pakistan and Iraq

In [0]:
%sql
SELECT location, EXTRACT(week from occurred_at) AS week, COUNT (*) AS engagement_count FROM events
WHERE event_type = 'engagement' AND occurred_at > '2014-06-30'
GROUP BY 1, 2
ORDER BY engagement_count DESC

location week engagement_count United States 27 2652 United States 31 2527 United States 30 2431 United States 29 2351 United States 28 2135 United States 33 1852 United States 32 1825 United States 34 1549 United States 35 1522 Japan 31 755 Japan 27 741 Germany 30 694 Germany 29 657 Germany 31 653 Japan 29 634 Germany 28 613 Japan 30 552 Japan 28 547 Japan 32 496 France 27 496 Japan 33 491 Japan 34 491 Germany 32 455 Germany 27 437 Russia 29 430 Germany 34 417 France 28 413 Germany 33 402 Italy 31 391 United Kingdom 35 378 Japan 35 376 France 31 369 Brazil 27 366 Russia 28 357 Germany 35 355 United Kingdom 31 337 France 30 332 United Kingdom 30 319 Italy 32 318 Russia 30 318 Italy 28 316 France 29 312 Russia 27 311 Brazil 28 297 Italy 27 297 United Kingdom 34 295 Russia 32 294 France 33 292 Russia 31 289 Russia 35 282 India 29 280 Korea 29 279 Italy 30 275 France 35 274 Russia 34 266 United Kingdom 28 263 Australia 32 261 Italy 29 259 United Kingdom 32 246 Australia 31 244 India 31 236 Australia 33 234 United Kingdom 29 230 Canada 31 228 Korea 31 225 Italy 34 225 Italy 35 224 France 32 221 Canada 29 219 Russia 33 217 Korea 27 216 Australia 28 216 France 34 215 Brazil 35 214 Brazil 31 214 Canada 28 213 Canada 32 212 United Kingdom 27 206 India 30 199 Brazil 30 198 Canada 34 193 Canada 30 188 United Kingdom 33 187 India 28 186 Brazil 34 185 Venezuela 30 185 India 27 184 Australia 34 182 Austria 28 179 Italy 33 176 Sweden 30 172 Brazil 33 172 Sweden 27 171 Mexico 27 170 Spain 35 167 Poland 31 160 Korea 32 160 Netherlands 34 158 India 32 158 Netherlands 31 158 India 35 154 Sweden 33 154 Switzerland 30 153 Poland 34 149 Korea 30 149 Australia 35 149 India 34 148 Mexico 35 148 Saudi Arabia 31 143 Egypt 35 142 Taiwan 27 140 Saudi Arabia 29 140 Saudi Arabia 28 138 Canada 35 135 Sweden 29 133 Brazil 29 130 Sweden 32 127 Brazil 32 126 Malaysia 28 124 Netherlands 32 124 India 33 121 Canada 27 121 Taiwan 33 120 Denmark 29 119 Finland 35 119 Switzerland 29 119 Mexico 29 118 Austria 29 118 Spain 34 118 Belgium 27 116 Mexico 33 114 Mexico 32 113 Mexico 28 113 Austria 27 112 Korea 33 111 Spain 28 111 Netherlands 33 109 Chile 32 109 Mexico 31 107 Chile 31 106 Taiwan 34 106 Spain 30 104 Iran 30 103 Sweden 34 102 Belgium 29 102 Belgium 30 101 Venezuela 29 101 Iran 32 101 Malaysia 31 101 Canada 33 101 Poland 29 100 Korea 35 100 Taiwan 32 100 Korea 28 98 Denmark 35 98 Indonesia 27 98 Iran 31 97 Spain 32 97 Saudi Arabia 30 97 South Africa 27 97 Belgium 28 96 Mexico 34 95 Netherlands 30 93 Australia 27 93 Australia 29 93 Norway 28 91 Mexico 30 90 Sweden 35 89 Colombia 29 88 Greece 29 88 Denmark 33 87 Iran 34 87 Turkey 32 87 Spain 31 86 Indonesia 35 85 Denmark 32 85 Indonesia 28 84 Israel 32 84 Saudi Arabia 32 82 Greece 28 82 Indonesia 29 81 Korea 34 80 Sweden 28 79 Chile 30 79 Egypt 34 79 Saudi Arabia 35 78 Malaysia 34 78 Nigeria 35 77 Netherlands 35 76 Belgium 31 76 Poland 35 76 Egypt 28 76 Spain 33 74 Switzerland 28 74 Austria 35 74 Poland 30 73 Turkey 28 72 Indonesia 34 72 Norway 27 71 Netherlands 28 71 Venezuela 31 70 Iran 27 70 Finland 34 69 Denmark 28 69 Singapore 29 69 Austria 31 68 Poland 27 68 Denmark 31 66 Pakistan 34 66 Spain 29 65 Saudi Arabia 33 65 Israel 33 65 Austria 34 64 Spain 27 64 Norway 32 62 Iran 29 62 Poland 32 62 Taiwan 28 62 Sweden 31 61 Colombia 31 61 Chile 29 59 Colombia 28 59 Denmark 27 59 Malaysia 29 59 Indonesia 33 59 Pakistan 35 58 Australia 30 58 Philippines 29 56 Iraq 31 56 Saudi Arabia 34 56 Egypt 29 56 Norway 29 55 Netherlands 27 55 Norway 33 54 Poland 33 54 Nigeria 29 53 South Africa 35 52 Poland 28 52 Venezuela 28 52 Switzerland 35 51 Malaysia 32 50 Norway 30 50 Taiwan 29 49 Malaysia 27 48 South Africa 34 48 Thailand 28 47 Philippines 28 47 South Africa 33 47 Nigeria 33 46 Greece 27 46 Singapore 27 46 Indonesia 31 45 Egypt 27 45 South Africa 28 45 Taiwan 30 45 Saudi Arabia 27 43 Chile 28 43 United Arab Emirates 31 43 Argentina 35 43 Turkey 30 42 Finland 33 41 Chile 35 41 Denmark 30 40 South Africa 31 40 Nigeria 2

##### Findings
- The major impact is the United States where the user engagement is the highest
- Switzerland had increased the no. of user actitivity learnierly from week 27 to 30 and then sudden decline after week 30 was seen which remained almost constant till week 35
- Neatherland and Spain had increased the number of user activities from 60 to 160 
- Austria had a linear decline in the user activity at week 28 from 180 users to 30 users
- Similarly Venezuela had also seen quite decline in the number of count in the user activity
- ALl the outher contries had somewhat up and downs in the activity but overall not caused much impact and remained stable

### Business Insights from the data set. 


- Now that we have analyzed drop of engagement in august but we don't know what was the reason behind this drop
- Drop of the user engagement activity can be due to various reason such as lost of interest of the users, New better subcription avaiable at loewr cost and more benefits
- Changes in the agreement with respect to the subcribers of a particular country

### Additional Task 
#### Do the answers to any of the initial questions lead you to additional questions or possibilities that should be investigated? If so, what are they and how will you address them?
##### ANSWER
- Yes the intial questions lead to additional questions and possibility that should be investigated such as whay few countries had so less user engagement
- Was there a drop in the user engagement during the weekends and rise in the engagement during the week to understand this we can make two graphs and see the user engagement on weekend Vs Weekdays
- We can also investigate in deep why the user activity dropped in the month of August and see what are the most affecting factors associated with this drop 
- We can also investigate the which days are affecting the drop in the user activities in the particular country (Might be because of public holidays)
- We can also investigate further if the user engagement depends on the culture or language of english speaking users in that particular country